### PR TITLE
bls test vectors

### PR DESCRIPTION
**Pull request by curiecrypt**
Test vectors for `bls12-381`, still under construction!

### CODE REVIEW

**Comment by lehins** [comment]
What's going on? Why are we getting PRs with Rust?

**Comment by iquerejeta** [comment]
So this is for the test vectors. We were thinking the other day with @tdammers were should the script that creates test vectors for BLS12-381 bindings live. The conclusion was that the best place for that code to live was in the repo where the test-vectors lived, hence cardano-base. However, this is obviously open to discussion if you think it should be outside of the repo. 

The motivation to put it here is that we could include a test in the CI that checks that the file used for the test vectors is equal to what is generated by the rust script. However, the idea is that we don't put any dependencies on the haskell build with rust. i.e., the rust script only needs to be run by the CI or anyone wanting to re-generate the test vectors. 

WDYT?

**Comment by iquerejeta** [comment]
I understand. The reason why we decided to do the test vectors in rust was to compare the library we use (`blst`) with zkcrypto's [library](https://github.com/zkcrypto/bls12_381) of the same curve, which is a library used in production. Creating the test vectors in C or Haskell would require bindings, and we'd depending on the rust library regardless, so IMO it doesn't really make sense. 

I understand wanting to leave rust outside of the build CI, in which case we can upload the rust code in another repo, link to that in the README and only have here the test vectors. 

If that's your preferred path, we'll go that way 👌

**Comment by lehins** [comment]
> Creating the test vectors in C or Haskell would require bindings

I don't understand why would we need any sort of bindings if we wanted to write a C program that generates test vectors. I would expect it to be same as the rust program in this PR, which depends on Rust implementation of bls, except that program would be written in C and depend on then C implementation bls

> I understand wanting to leave rust outside of the build CI, in which case we can upload the rust code in another repo, link to that in the README and only have here the test vectors.

If you think what I am suggesting is too hard to do then we don't need to waste any time and I am not gonna stand in way of having this code in this repo. After all, it will only be used once to generate the test vectors. If we had to run it on CI and locally every time we had to run the test suite that would be a different story and I would be pushing back much harder against it.

**Comment by lehins** [comment]
> my intention was to include a check in CI that the test-vectors file used for tests corresponds to the output of the rust binary

The output of the Rust program will be checked-in in this repo, no? Then why do we need to run Rust on CI?

> But we want to create these test vectors with the rust library, which is the 'other' reference implementation of bls12-381. We could do it with the underlying library, but the goal of these test vectors is that they are independently generated.

Which library are you trying to test? Are you trying to test the `libblst` library implementation or do you want to test the Haskell binding to `liblst` that we provide in `cardano-crypto-class` library? If it is the latter, then there is nothing wrong with using the same C library to generate test vectors. However, if it is the former and you want to test that Rust implementation matches the C implementation, then such test suite has no business being being in this repository as it is unrelated to our Haskell bindings in any way. A test like this would belong in either the Rust library or in the C library repository.

My understanding was that the reason for writing this script in Rust was only because it is easier to write Rust, rather than C.

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/Cargo.toml
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/Cargo.toml:11]: Not needed (`rand_core` imported from `rand_chacha`)

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/Cargo.toml
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/Cargo.toml:17]: Ditto

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/Cargo.toml
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/Cargo.toml:7]: No need to keep this

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/README.md
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/README.md:17]: Missing introduction on what operations can be tested with these test vectors (addition, subtraction, scalar mult and negation)

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/README.md
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/README.md:30]: Is the scalar stored somewhere?

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/README.md
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/README.md:47]: I would change the name of these variables from `G1_uncom_curve` into `G1_uncom_not_on_curve`. Also, The question after the semicolon does not seem to provide any extra info.

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/README.md
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/README.md:50]: A little more context needed here. In the cardano-base bindings we don't allow the user to use `DST` and `aug` explicitly, and these need to be appended into the, by hand (following [hash-to-curve](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve#name-expand_message) spec), message before verifying.

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/bls_sig_aug_test_vectors
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/bls_sig_aug_test_vectors:1]: Include all test-vector files in a different folder.

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/src/main.rs
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/src/main.rs:18]: You can operate in Projective, and do the conversion just at the end.

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/src/main.rs
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/src/main.rs:55]: You could create a helper function to take as input a slice of hex strings. That would simplify the code of this script.

File: cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/src/main.rs
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/test-vectors-bls12_381/src/main.rs:164]: This should happen inside the loop, otherwise, it could happen that no point is found and we store a slice of zeroes.

**Review by iquerejeta — CHANGES_REQUESTED** [review]
Left a few comments. Some of the are generic that can be applied in several spots of the script (e.g. the usage of projective point, or the usage of a helper function). Two final comments:
* I'd suggest we put the data files in a different folder
* We now have `bls12-381-test-vectors` that only contains `test-vectors-bls12_381`. I'd suggest we put the contents of `test-vectors-bls12_381` directly in the first folder.

**Review by curiecrypt — COMMENTED** [review]
LGTM

**Comment by lehins** [comment]
@iquerejeta This PR is up your alley. We've already talked about this work, so if you can oversee this work further that'll be great.

I suspect Haskell tests will be added in a separate PR, right?

**Comment by iquerejeta** [comment]
Haskell tests will be added in a separate PR, yes. I will oversee the PR. If you could approve it to unblock your change request, so that once I approve we can merge.

File: cardano-crypto-tests/bls12-381-test-vectors/README.md
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/README.md:86]: We use zkcrypto to create the vectors, but guarantee that they are valid wrt to the `blst` bindings:
```suggestion
Test vectors for BLS signature, with successful validation using [blst](https://github.com/supranational/blst).
```

File: cardano-crypto-tests/bls12-381-test-vectors/README.md
Comment (iquerejeta) [inline cardano-crypto-tests/bls12-381-test-vectors/README.md:95]:
```suggestion
let aug = b"Random value for test aug. ";
```

**Review by iquerejeta — APPROVED** [review]
Looks good. Some minor suggestions before merging.

**Pull request merged**
